File: apps/go/productivity_tool/occurrence_tracker/alert_handler.go
package main

import (
	"fmt"
	"github.com/avilikof/monorepo/lib/go/alert_entity"
	"log"
	"strconv"
	"time"
)

type OccurrenceHandler struct {
	internalStorage map[string][]byte
}

func NewOccurrenceHandler() OccurrenceHandler {
	return OccurrenceHandler{internalStorage: make(map[string][]byte)}
}

func (h *OccurrenceHandler) Handle(alert alert_entity.AlertEntity) error {
	if h.alertExistsInStorage(&alert) {
		return h.existingAlert(&alert)
	}
	if alert.GetState() == "firing" {
		return h.newOccurrence(false, &alert)
	}
	return nil
}
func (h *OccurrenceHandler) existingAlert(alert *alert_entity.AlertEntity) error {
	oldAlert, _err := h.getAlertFromStorage(alert)
	if _err != nil {
		return _err
	}
	if alert.GetState() == "firing" {
		if oldAlert.GetState() == "firing" {
			return nil
		}
		return h.newOccurrence(true, alert)
	}
	if oldAlert.GetState() == "resolved" {
		return nil
	}
	return h.resolve(alert)
}
func (h *OccurrenceHandler) getAlertFromStorage(alert *alert_entity.AlertEntity) (alert_entity.AlertEntity, error) {
	if h.alertExistsInStorage(alert) {
		oldAlertBytes, ok := h.internalStorage[alert.GetOccurrenceId()]
		if !ok {
			return alert_entity.AlertEntity{}, fmt.Errorf("failed to get alert from internal storage")
		}
		oldAlert, _err := alert_entity.NewAlertEntityFromBytes(oldAlertBytes)
		if _err != nil {
			return alert_entity.AlertEntity{}, _err
		}
		return *oldAlert, nil
	}
	return alert_entity.AlertEntity{}, fmt.Errorf("alert not found in internal storage")
}

func (h *OccurrenceHandler) alertExistsInStorage(alert *alert_entity.AlertEntity) bool {
	_, ok := h.internalStorage[alert.GetAlertId()]
	return ok
}

func (h *OccurrenceHandler) pushToStorage(alert alert_entity.AlertEntity) error {
	alertBytes, _err := alert.ToByte()
	if _err != nil {
		return _err
	}
	log.Println(alert)
	h.internalStorage[alert.GetAlertId()] = alertBytes
	return nil
}

func (h *OccurrenceHandler) newOccurrence(reopen bool, alert *alert_entity.AlertEntity) error {
	if reopen {
		_err := alert.SetDescription(alert.GetDescription() + " reopen ")
		if _err != nil {
			return _err
		}
	}
	_err := alert.SetOccurrenceId(strconv.FormatInt(time.Now().UnixNano(), 10))
	if _err != nil {
		return _err
	}
	return h.pushToStorage(*alert)
}

func (h *OccurrenceHandler) resolve(alert *alert_entity.AlertEntity) error {
	oldAlert, _err := h.getAlertFromStorage(alert)
	if _err != nil {
		return _err
	}
	_err = oldAlert.SetState("resolved")
	if _err != nil {
		return _err
	}
	return h.pushToStorage(oldAlert)
}

File: apps/rust/productivity_tool/occurrency_tracker/src/handlers/alert_handler.rs
use crate::interfaces::repo_interface::RepoInterface;
use alert_entity::{AlertEntity, AlertState};
use event_entity::{EventAction, EventEntity, EventType};
use log::{debug, error};

pub struct AlertHandler<'a, R>
where
    R: RepoInterface,
{
    received_alert: AlertEntity,
    repo: &'a mut R,
    service: String,
}

impl<'a, R> AlertHandler<'a, R>
where
    R: RepoInterface,
{
    pub fn init(received_alert: &AlertEntity, repo: &'a mut R) -> Self {
        Self {
            received_alert: received_alert.to_owned(),
            repo,
            service: "occurrence-handler".to_string(),
        }
    }
    pub fn occurrence_handling_flow(&mut self) -> EventEntity {
        //     First step - check if alert with same alert_id exists in storage
        match self.extract_alert_from_repo() {
            None => self.new_occurrence(),
            Some(_) => self.handle_existing_alert(),
        }
    }
    fn new_occurrence(&mut self) -> EventEntity {
        const DROP_DESCRIPTION: &str = "dropping new resolved alert";
        match self.received_alert.get_state() {
            AlertState::Firing => self.first_occurrence(),
            AlertState::Resolved => {
                debug!("{DROP_DESCRIPTION}");
                event_action_drop(
                    &mut self.received_alert,
                    DROP_DESCRIPTION,
                    self.service.as_str(),
                )
            }
        }
    }
    fn handle_existing_alert(&mut self) -> EventEntity {
        const ERROR_MESSAGE: &str = "expected alert to exists in repo but it doesnt";
        match self.extract_alert_from_repo() {
            None => {
                error!("{ERROR_MESSAGE}");
                EventEntity::new(
                    &mut self.received_alert,
                    EventType::Log,
                    EventAction::Failure,
                    ERROR_MESSAGE,
                    self.service.as_str(),
                )
            }
            Some(alert_from_repo) => match alert_from_repo.get_state() {
                AlertState::Firing => self.existing_alert_firing(),
                AlertState::Resolved => self.existing_alert_resolved(),
            },
        }
    }
    fn first_occurrence(&mut self) -> EventEntity {
        const DESCRIPTION: &str = "alert opened";
        let mut new_alert = self.received_alert.clone();
        new_alert.set_new_occurrence_id();
        let new_alert_bytes = new_alert.as_bytes().unwrap();
        self.repo
            .push(new_alert.get_alert_id().to_string(), new_alert_bytes);
        debug!("{DESCRIPTION}");
        EventEntity::new(
            &mut new_alert,
            EventType::Event,
            EventAction::Open,
            DESCRIPTION,
            self.service.as_str(),
        )
    }
    fn existing_alert_firing(&mut self) -> EventEntity {
        const ERROR_DESCRIPTION: &str =
            "old alert has no `occurrence_id` but is expected to have one";
        const REFIRING: &str = "alert refiring";
        match self.received_alert.get_state() {
            AlertState::Firing => {
                let mut old_alert = self.get_existing_alert_from_repo();
                match old_alert.get_occurrence_id() {
                    None => {
                        event_type_log(&mut old_alert, ERROR_DESCRIPTION, self.service.as_str())
                    }
                    Some(id) => {
                        self.received_alert.set_occurrence_id(id.to_string());
                        event_action_drop(&mut self.received_alert, REFIRING, self.service.as_str())
                    }
                }
            }
            AlertState::Resolved => self.resolve(),
        }
    }
    fn existing_alert_resolved(&mut self) -> EventEntity {
        match self.received_alert.get_state() {
            AlertState::Firing => self.reopen(),
            AlertState::Resolved => event_action_drop(
                &mut self.received_alert,
                "alert dropped id 1",
                &self.service,
            ),
        }
    }
    fn reopen(&mut self) -> EventEntity {
        const REOPEN: &str = "alert reopened";
        self.received_alert.set_new_occurrence_id();
        let new_alert = self.received_alert.clone().as_bytes().unwrap();
        self.repo
            .update(self.received_alert.get_alert_id().to_string(), new_alert);
        EventEntity::new(
            &mut self.received_alert,
            EventType::Event,
            EventAction::Reopen,
            REOPEN,
            self.service.as_str(),
        )
    }
    fn resolve(&mut self) -> EventEntity {
        const ERROR_ALERT_NOT_IN_REPO: &str = "failure, alert not found in repo";
        const ERROR_DESERIALIZING: &str = "failed to deserialize alert";
        match self.repo.pull(self.received_alert.get_alert_id()) {
            None => {
                error!("{ERROR_ALERT_NOT_IN_REPO}");
                event_type_log(
                    &mut self.received_alert,
                    ERROR_DESERIALIZING,
                    self.service.as_str(),
                )
            }
            Some(b) => match AlertEntity::from_bytes(b.as_slice()) {
                Ok(new_alert) => self.resolve_alert(new_alert),
                Err(err) => {
                    let err_message = format!("{}: {}", ERROR_DESERIALIZING, err);
                    error!("{}", err_message);
                    event_type_log(
                        &mut self.received_alert,
                        err_message.as_str(),
                        self.service.as_str(),
                    )
                }
            },
        }
    }
    fn resolve_alert(&mut self, mut new_alert: AlertEntity) -> EventEntity {
        const RESOLVED: &str = "alert resolved";

        new_alert.set_state(AlertState::Resolved);
        let alert_bytes = new_alert.as_bytes().unwrap();
        self.repo
            .update(new_alert.get_alert_id().to_string(), alert_bytes);
        debug!("resolved");
        EventEntity::new(
            &mut new_alert,
            EventType::Event,
            EventAction::Resolve,
            RESOLVED,
            self.service.as_str(),
        )
    }
    fn extract_alert_from_repo(&self) -> Option<AlertEntity> {
        match self.repo.pull(self.received_alert.get_alert_id()) {
            None => None,
            Some(payload) => match AlertEntity::from_bytes(payload) {
                Err(_) => None,
                Ok(a) => Some(a),
            },
        }
    }
    fn get_existing_alert_from_repo(&self) -> AlertEntity {
        match self.repo.pull(self.received_alert.get_alert_id()) {
            Some(alert) => AlertEntity::from_bytes(alert).unwrap(),
            _ => panic!("something really wrong"),
        }
    }
}

fn event_action_drop(alert: &mut AlertEntity, description: &str, name: &str) -> EventEntity {
    EventEntity::new(
        alert,
        EventType::Event,
        EventAction::Drop,
        description,
        name,
    )
}

fn event_type_log(alert: &mut AlertEntity, description: &str, name: &str) -> EventEntity {
    EventEntity::new(
        alert,
        EventType::Log,
        EventAction::Failure,
        description,
        name,
    )
}
